### PR TITLE
Enable/disable checking of task exit code to indicate task failure

### PR DESCRIPTION
**Pull request by kshakir**
DSDEEPB-734

### CODE REVIEW

File: src/main/java/cromwell/parser/RuntimeKey.java [hunk 1]
```diff
@@ -11,7 +12,8 @@
     DEFAULT_DISKS("defaultDisks", JES),
     DEFAULT_ZONES("defaultZones", JES),
     DOCKER("docker", new BackendType[]{JES}, LOCAL), // Alternate constructor due to both optional and mandatory backends
-    FAIL_ON_STDERR("failOnStderr", LOCAL, JES),
+    FAIL_ON_STDERR("failOnStderr", JES, LOCAL, SGE),
```
Comment (mcovarr) [inline src/main/java/cromwell/parser/RuntimeKey.java]: Was the change to add JES for `failOnStderr` intentional?

File: src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala [hunk 2]
```diff
@@ -140,7 +140,8 @@ class SgeBackend extends Backend with SharedFileSystem with LazyLogging {
     logger.info(s"$tag SGE job completed (rc=$jobRc)")
     (jobRc, backendCall.stderr.toFile.length) match {
       case (r, _) if r == 143 => AbortedExecution // Special case to check for SIGTERM exit code - implying abort
-      case (r, _) if r != 0 => FailedExecution(new Throwable(s"$tag SGE job failed because of non-zero return code: $r"), Option(r))
+      case (r, _) if r != 0 && backendCall.call.failOnRc =>
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala]: Unless the requirements have changed, the ticket specifies a whitelist of non-failing rc's.

File: src/main/scala/cromwell/engine/backend/local/LocalBackend.scala [hunk 3]
```diff
@@ -165,16 +165,21 @@ class LocalBackend extends Backend with SharedFileSystem with LazyLogging {
...
-            case Success(outputs) => SuccessfulExecution(outputs)
-            case Failure(e) =>
-              val message = Option(e.getMessage) map { ": " + _ } getOrElse ""
-              FailedExecution(new Throwable("Failed post processing of outputs" + message, e))
-          }
+        case Success(0) => processSuccess()
         case Success(143) => AbortedExecution // Special case to check for SIGTERM exit code - implying abort
-        case Success(badRc) => FailedExecution(new Throwable(s"Call ${backendCall.call.fullyQualifiedName}, Workflow ${backendCall.workflowDescriptor.id}: $rc\n\nFull command was: ${argv.mkString(" ")}"), Option(badRc))
+        case Success(badRc) if !backendCall.call.failOnRc => processSuccess()
+        case Success(badRc) if backendCall.call.failOnRc => FailedExecution(new Throwable(s"Call ${backendCall.call.fullyQualifiedName}, Workflow ${backendCall.workflowDescriptor.id}: $rc\n\nFull command was: ${argv.mkString(" ")}"), Option(badRc))
```
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/local/LocalBackend.scala]: Do we need `backendCall.call.failOnRc` if we just tested the negation on the preceding line?

**Comment by scottfrazer** [comment]
can both `failOnStderr=false` and `failOnRc=false`?

File: src/test/scala/cromwell/util/SampleWdl.scala [hunk 4]
```diff
@@ -1220,4 +1220,33 @@ object SampleWdl {
...
+        |  }
+        |}
+        |
+        |
+        |workflow w {
+        |  call A
+        |  call B {input: B_in=A.A_out}
+        |}
+      """.stripMargin.replaceAll("RUNTIME", runtime)
+    override lazy val rawInputs = Map("" -> "...")
```
Comment (mcovarr) [inline src/test/scala/cromwell/util/SampleWdl.scala]: huh?  why not `Map.empty`?

Comment (scottfrazer) [inline src/test/scala/cromwell/util/SampleWdl.scala]: Probably copy/paste... I've seen this in other places in this file.

File: src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala (hunk 2 above)
Comment (geoffjentry) [inline src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala]: To be pedantic (cuz I roll like that) the ticket says non-zero exit code. @scottfrazer asked if we could do it as a whitelist of codes and the author of the ticket said that was AOK.

File: src/main/scala/cromwell/binding/RuntimeAttributes.scala [hunk 5]
```diff
@@ -79,6 +82,10 @@ object RuntimeAttributes {
     value map validateBoolean getOrElse Defaults.FailOnStderr.successNel
   }
 
+  private def validateFailOnRc(value: Option[String]): ValidationNel[String, Boolean] = {
+    value map validateBoolean getOrElse Defaults.FailOnRc.successNel
```
Comment (cjllanwarne) [inline src/main/scala/cromwell/binding/RuntimeAttributes.scala]: (Probably misreading this), why are these all returning success if the validation fails?

Comment (geoffjentry) [inline src/main/scala/cromwell/binding/RuntimeAttributes.scala]: The `getOrElse` is handling the case where nothing was passed in, so it's just the `SuccessNel` of our default value

File: src/test/scala/cromwell/FailOnRcSpec.scala [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        sampleWdl = SampleWdl.FailOnRc,
+        runtime = "runtime {failOnRc: true}",
+        eventFilter = EventFilter.info(pattern = s"persisting status of w.A to Failed", occurrences = 1),
+        stdout = Map("w.A" -> Seq("321\n")),
+        stderr = Map("w.A" -> Seq("")),
+        terminalState = WorkflowFailed
+      )
+    }
+
+    "have correct contents in stdout/stderr files for a call that does fail on rc" in {
```
Comment (Horneth) [inline src/test/scala/cromwell/FailOnRcSpec.scala]: `that does *not* fail on rc` ?

File: src/main/java/cromwell/parser/RuntimeKey.java (hunk 1 above)
Comment (kshakir) [inline src/main/java/cromwell/parser/RuntimeKey.java]: Explicitly added SGE. OCD-ish tendencies made me put JES first.

Comment (mcovarr) [inline src/main/java/cromwell/parser/RuntimeKey.java]: whoops ok

File: src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala (hunk 2 above)
Comment (mcovarr) [inline src/main/scala/cromwell/engine/backend/sge/SgeBackend.scala]: Per discussion with Jeff, the original ticket did say non-zero exit code so presumably that's enough to be of value to the author of the story.  We can handle the whitelist as a separate future enhancement.

**Comment by kshakir** [comment]
Verbal "ok" received from various members for now. Merging.

**Pull request merged**